### ISSUE TITLE
Can't persist changes to an entity when relationships exist

### ISSUE THREAD

**Issue by JoeThielen** [graphaware/neo4j-php-ogm#64]
I'm following the demo, with great success, but I've come to a halt and I have no idea what the issue is.

I've come to this part, but can't get it to work:

```

// Find Tom Hanks, filter his movies to find Cast Away and rename it to Cast Away 2
/** @var Person $tomHanks */
$tomHanks = $manager->getRepository(Person::class)->findOneBy('name', 'Tom Hanks');
$filter = array_values(array_filter($tomHanks->getMovies()->toArray(), function(\Movies\Movie $movie) {
    return 'Cast Away' === $movie->getTitle();
}));

/** @var \Movies\Movie $castAway */
$castAway = $filter[0];
$castAway->setTitle('Cast Away 2');
$manager->flush();

```

If I echo $castAway->getTitle() I sure enough get "Cast Away 2", but the save is not actually persisted in the DB.  

I've tried this on my own with both Movies and Persons and here are my findings:
- When a movie or person has NO relationships, the change WORKS.
- When a movie or person has relationships, the change DOES NOT WORK.

The original test code in  /docs/_code/movies/Movie.php seems to have worked because of an incorrect directional mapping (OUTGOING vs INCOMING), which meant relationships were not actually populating.  But when I make that change and they do populate, then the code stops working.

I don't know enough to know where to begin looking on this one.  I did note that Movie.php does not have a mappedBy parameter on the relationship annotation for $actors, but when I put that in it didn't affect anything.

Any clues here?

**Comment by theUm** [comment]
It seems that this project has major internal mechanics difficulties. I'm using it almost 2 months now, and I ended up writing ALL create/update queries by myself. I have lot of circular references and some-to-some relations and if I try to manage them by entityManager they loosing their links in one case and they are formed into separate duplicated node tree with same nodes data...
Anyway I want to use this orm as DB data mapper and Object accessor with lazy load. Aaand, guess what, because of circular reference or some misterious reason I cant access third level many-to-one relationship node data, like this `$object->getChilds()->first()->getObjectType()`. `$object` has many childs that have exactly one `$objectType`. I can never get `$objectType` like that.
So regarded this ORM I have very mixed feelings - on the one hand project team has written big smart ORM which in theory could fill empty space between Neo4j  and PHP. On the other hand if you have enough time to dig up ORM behaviour and code you will see some really ~~fuc~~ strange things, like copying of your entire node tree onto separate tree

**Comment by ikwattro** [comment]
Yup there hasn't been a lot of updates since 2-3 months I agree, mainly due to international move and work load. I start to have some free time again to look at it.

**Comment by theUm** [comment]
It`s really good news to hear. ORM is really promising

**Comment by ikwattro** [comment]
Hi @JoeThielen , @theUm 

I did a serious progress this night on the refactoring of the hydration part, everything is in the rc1 branch. Code is smaller and simpler, all tests for those are passing. Some changes occured for the findBy() but will document it (most of the changes are migration do Doctrine like methods so will ease usage for people knowing already Doctrine)

I will make a full test on the movies dataset and adapt the doc.

If you will help @theUm , you can create a test in https://github.com/graphaware/neo4j-php-ogm/tree/rc1/tests/Community for your issue and add your example Model , this would be very lovely. Normally circular references shouldn't happen anymore, normally :)

**Comment by ikwattro** [comment]
Update: Merged in master. Create a branch alias for 1.0 to master

**Comment by ikwattro** [comment]
@JoeThielen @theUm 

Hey guys,

I confirm the issue is fixed per the test in https://github.com/graphaware/neo4j-php-ogm/blob/master/tests/Integration/MovieDatasetTest.php#L48

Note on some current BC : 

Due to the new Proxy'zation of entities (like in Doctrine but not hardcore yet), accessing collections like this : 

```php
$actors = $this->actors;
```

will not trigger the initialization and the lazy loading from the database, only fetching via getters are supported for now : 

```php
$actors = $this->getActors();
```

This will be fixed later on, but I prioritize a complete working flow before this.

Cheers !

**Comment by JoeThielen** [comment]
👍

Yes!  I just did a very quick test after upgrading to 1.0.0-RC1 and changing the findByOne() parameters to arrays (that will definitely need changed in the docs and example code if not done already...!) and it did work.  Thanks, promising.

**Comment by ikwattro** [comment]
@JoeThielen Nice to hear, docs are updated already ;-)

**Comment by JoeThielen** [comment]
Is there an issue opened for this:

> 
> Note on some current BC :
> 
> Due to the new Proxy'zation of entities (like in Doctrine but not hardcore yet), accessing collections like this :
> 
> $actors = $this->actors;
> 
> will not trigger the initialization and the lazy loading from the database, only fetching via getters are supported for now :
> 
> $actors = $this->getActors();
> 
> This will be fixed later on, but I prioritize a complete working flow before this.

I noted during my testing that the User example portions of the docs does work as explained when creating a new user, but when looking up an existing user, it fails due to what I believe is a lazy loading issue.  I apologize as all this is pretty new to me, so if I'm barking up the wrong tree please forgive me.  

This works:
```
$user = new JTUser('cypher666');
$user->rateMovie($movie, '4.5');
$manager->persist($user);
$manager->flush();
```

This does not:
```
$user = $userRepository->findOneBy(array('login' => 'cypher666'));
$movie = $manager->getRepository(JTMovie::class)->findOneBy(array('title' => 'The Matrix Reloaded'));
$user->rateMovie($movie, '4.5');
$manager->persist($user);
$manager->flush();
```

I get this error:

> PHP Fatal error:  Uncaught Error: Call to a member function add() on null ...

But if I edit the rateMovie() method of the User class and add this at the top of that method, then it does work.

`$this->getRatings();
`

**Thank you** for noting this behavior in your comment, otherwise I would not have had any idea what was going on. 

Just curious if this was scheduled to be fixed for the 1.0 release, I didn't quite see anything covering this in the issues although you did say it would fixed "later on".

**Comment by ikwattro** [comment]
Interesting, let me add a test for this. RC-2 might already come by the end of the week though :)

**Comment by ikwattro** [comment]
Sorry yes I changed the doc but not the code in the doc. Good catch @JoeThielen . Thanks !

This will be fixed for sure before the 1.0.0 release, so during the release candidate phase.

**Comment by CristianSitov** [comment]
Is this issue resurfacing?

I have several LegalDocument() entities that belong to a Resource() entity (one-to-many relationship). If iterate and try to edit each LegalDocument() by setting its properties then it works ok. If I try to removeElement() (one or multiple) LegalDocument, it works fine. But, if I have several LegalDocument() attached to Resource() and I want to add one more then it breaks all relationships and I end up with all LegalDocument() unattached and a new LegalDocument() freshly created. The add() methods breaks everything.
```
$newLegalDocument = new LegalDocument();
$resource->getLegalDocuments()->add($newLegalDocument);
```
